### PR TITLE
[Fleet] Add e2e tests for api versions

### LINKED ISSUE #2594: Fleet Server testing for versioned API
Fleet server APIs have to be versioned in order to ensure backwards compatibility with all already existing components calling it.
This issue is a placeholder that will be used as soon as the first versioned API is introduced.

```[tasklist]
### Tasks
- [ ] /api/status:
- [ ] /api/fleet/agents/enroll:
- [ ] /api/fleet/agents/{id}/checkin:
- [ ] /api/fleet/agents/{id}/acks:
- [ ] /api/fleet/artifacts/{id}/{sha2}:
- [ ] /api/fleet/uploads:
- [ ] /api/fleet/uploads/{id}/{chunkNum}:
- [ ] /api/fleet/uploads/{id}
```

### PR DESCRIPTION
**Pull request by nchaulet**
## Description

Resolve #2594

That PR add e2e test for versioned APIs.

To do so I duplicated the client for each API version, and created a duplicated test suite for each version.

We probably want to relax our sonar rules for duplication in that case.

### CODE REVIEW

File: testing/e2e/stand_alone_test.go [hunk 1]
```diff
@@ -155,94 +156,91 @@ func (suite *StandAloneSuite) TestClientAPI() {
...
-	// Run subtests here
-	suite.Run("test status unauthenicated", func() {
-		ctx, cancel := context.WithCancel(bCtx)
-		defer cancel()
-		tester := &ClientAPITester{
-			suite.Suite,
-			ctx,
-			suite.client,
-			"https://localhost:8220",
+	versions := []string{"current", "2022-06-01"}
```
Comment (nchaulet) [inline testing/e2e/stand_alone_test.go:159]: Run the test for each the version here, if we add a new version we will have to create a new `api_version/client_api*.go` (probably worth adding this to the readme)

**Comment by jlind23** [comment]
Nice job, I can also see all tests passing on the buildkite build:
Screenshot 2023-06-13 at 07 29 04

**Review by jlind23 — APPROVED** [review]

**Comment by jlind23** [comment]
NIT: Can we run the API tests in parallel like one "runner" per API version? Otherwise it might increase the build time.

File: testing/e2e/api_version/client_api_2023_06_01.go [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+		if withRequestEditorFn != nil {
+			return withRequestEditorFn(ctx, req)
+		}
+
+		return nil
+	}))
+}
+
+// TestStatus tests the status endpoint, if an apiKey is given the authenticated response if verfied.
+func (tester *ClientAPITester20230601) TestStatus(apiKey string) {
```
Comment (jsoriano) [inline testing/e2e/api_version/client_api_2023_06_01.go:57]: Are we going to duplicate all tests for all versions of the API?

File: testing/e2e/api_version/client_api_2023_06_01.go [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+	} else {
+		resp, err = client.StatusWithResponse(tester.ctx, &api.StatusParams{})
+	}
+	tester.Require().NoError(err)
+	tester.Require().Equal(http.StatusOK, resp.StatusCode())
+	tester.Require().NotEmpty(resp.JSON200.Name)
+	tester.Require().NotEmpty(resp.JSON200.Status)
+	if apiKey != "" {
+		tester.Require().NotNil(resp.JSON200.Version)
+	}
```
Comment (jsoriano) [inline testing/e2e/api_version/client_api_2023_06_01.go:76]: Nit. Use Assert for non-critical assertions so they are all executed.
```suggestion
	tester.Assert().Equal(http.StatusOK, resp.StatusCode())
	tester.Assert().NotEmpty(resp.JSON200.Name)
	tester.Assert().NotEmpty(resp.JSON200.Status)
	if apiKey != "" {
		tester.Assert().NotNil(resp.JSON200.Version)
	}
```

**Review by jsoriano — APPROVED** [review]

File: testing/e2e/api_version/api_version.go [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+// you may not use this file except in compliance with the Elastic License.
+
+package api_version
+
+// enrollMetadataTpl is the template for enrollement metadata
+// It defines elastic.agent.version as that attribute is required by Kibana to successfully request diagnostics from an agent.
+// A diagnostics request is used when testing the file-upload endpoints.
+const enrollMetadataTpl = `{"elastic":{"agent":{"version":"%s"}}}`
+
+type ClientAPITesterInterface interface {
```
[the remark this answers is not in the document]
Comment (nchaulet) [reply testing/e2e/api_version/api_version.go:12]: Looks like if I put in a `_test` file I cannot access it in the parent module

File: testing/e2e/api_version/client_api_2023_06_01.go (hunk 2 above)
Comment (nchaulet) [reply testing/e2e/api_version/client_api_2023_06_01.go:57]: Yes I think it's the easiest way to maintain a test suite for each version as normally we will have difference between version, but I am open to other suggestions

**Comment by jlind23** [comment]
@michel-laterman then we should probably keep an eye on this as it might become pretty slow when number of api version will increase.

File: testing/e2e/api_version/api_version.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (jsoriano) [reply testing/e2e/api_version/api_version.go:12]: Ah right, this is used in the parent. Well, everything is under `testing`, so it should be fine :slightly_smiling_face:

[the remark this answers is not in the document]
Comment (jsoriano) [reply testing/e2e/api_version/api_version.go:12]: But well, given that this is only used in the parent, I think we should define it there too.

**Comment by jlind23** [comment]
Then @nchaulet should be good to merge this one then.

**Pull request merged**